### PR TITLE
doc(fivetran_sdk): Updated development guide for partners

### PR DESCRIPTION
**Pull request by fivetran-satvikpatil**
Closes https://fivetran.height.app/T-711151

### Description of Changes
Updated development guide for partners

### CODE REVIEW

File: development-guide.md [hunk 1]
```diff
@@ -61,13 +61,32 @@ The following are hard requirements to be able to deploy Partner code to Fivetra
...
+## Source Connector Guidelines
 
 - Don't push anything other than source data to the destination. State will be saved to production DB and returned in `UpdateRequest`.
 - Don't forget to handle new schemas/tables/columns per the information and user choices in `UpdateRequest#selection`.
 - Make sure you checkpoint at least once an hour. The more frequently you do it, the better.
 
-## Destination Guidelines
+### RPC Calls
+#### Schema
+This operation should retrieve all the information about the customer's schemas, tables, and columns. It also offers an optional `selection_not_supported` field that indicates whether customers can select or deselect tables and columns within the Fivetran dashboard.
```
Comment (manjutapali) [inline development-guide.md:84]: This operation retrieves of customer's schemas, tables, and columns. It also includes an optional `selection_not_supported` field that indicates whether customers can select or deselect tables and columns within the Fivetran dashboard.

**Review by manjutapali — COMMENTED** [review]
@fivetran/tech-writers  Need to reformat the statements.

File: development-guide.md [hunk 2]
```diff
@@ -61,13 +61,32 @@ The following are hard requirements to be able to deploy Partner code to Fivetra
...
 
-## Connector Guidelines
+## Setup Form Guidelines
+- Keep the form clear and concise, only requesting essential information for successful connector setup.
+- Use clear and descriptive labels for each form field. Avoid technical jargon if possible.
+- Organize the fields in a logical order that reflects the setup process.
+
+### RPC Calls
+#### ConfigurationForm
+This operation retrieves all the setup form fields and tests info required to establish a connection. You can provide various parameters for the fields to enhance the user experience, such as descriptions, optional fields, and more.
```
[the remark this answers is not in the document]
Comment (fivetran-satvikpatil) [reply development-guide.md:71]: > Is this statement needed - `tests info required to establish a connection`.

The configurationForm retrieves all the setup form fields and tests. 
Should I just change the sentence to `This operation retrieves all the setup form fields and tests` ?

File: development-guide.md [hunk 3]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
 - This operation might be requested for a table that does not exist in the destination. In that case, it should NOT fail, simply ignore the request and return `success = true`.
 - `utc_delete_before` has millisecond precision.
 
-#### WriteBatchRequest
+#### WriteBatch
+This operations provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message::
```
Comment (5tran-alexil) [inline development-guide.md:128]:
```suggestion
This operation provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

File: development-guide.md [hunk 4]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+#### Capabilities
+This operation offers several additional features, as listed below:
+
+- Datatype Mappings: Supports adjusting partner data types for each Fivetran data type.
+- Max value for columns: Provides an option to specify the maximum value for each data type.
+
+#### AlterTable
+This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides details about these updates:
+- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in destination table.
+- Update Column type (`change_column_type`): This field provides the info on updated Column from source that needs to be reflected in destination table.
```
Comment (5tran-alexil) [inline development-guide.md:156]:
```suggestion
- Update Column type (`change_column_type`): This field provides information on updated type of a column in the source that needs to be reflected in a destination table.
```

**Review by 5tran-alexil — COMMENTED** [review]
See my suggestions

File: development-guide.md [hunk 5]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+- Primary key updates (`new_primary_key`):  If the customer's primary key have changed, this field lists all the new primary key used.
```
[the remark this answers is not in the document]
Comment (fivetran-satvikpatil) [reply development-guide.md:157]: @5tran-alexil What about this?

`Primary key updates (`new_primary_key`):  If the primary key has changed, this field lists all the new columns used in the revised primary key.`

[the remark this answers is not in the document]
Comment (5tran-alexil) [reply development-guide.md:157]:
```suggestion
- Primary key updates (`new_primary_key`):  If the primary key have changed, this field lists all the new columns used in the updated primary key.
```

[the remark this answers is not in the document]
Comment (5tran-alexil) [reply development-guide.md:157]: @fivetran-satvikpatil see ^^

File: development-guide.md [hunk 6]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+This operation offers several additional features, as listed below:
+
+- Datatype Mappings: Supports adjusting partner data types for each Fivetran data type.
+- Max value for columns: Provides an option to specify the maximum value for each data type.
+
+#### AlterTable
+This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides details about these updates:
+- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in destination table.
+- Update Column type (`change_column_type`): This field provides information on updated type of a column in the source that needs to be reflected in a destination table.
+- Primary key updates (`new_primary_key`):  If the primary key has changed, this field lists all the new columns used in the updated primary key.
```
Comment (fivetran-abdulsalam) [inline development-guide.md:157]:
```suggestion
- Primary key updates (`updated_primary_keys`):  If the primary key has changed, this field lists all the new columns used in the updated primary key.
```

File: development-guide.md [hunk 7]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
 - This operation might be requested for a table that does not exist in the destination. In that case, it should NOT fail, simply ignore the request and return `success = true`.
 - `utc_delete_before` has millisecond precision.
 
-#### WriteBatchRequest
+#### WriteBatch
+This operation provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```
Comment (5tran-alexil) [inline development-guide.md:128]:
```suggestion
This operation provides details about the batch files containing the retrieved data. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

File: development-guide.md [hunk 8]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
 - `replace_files` is for `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the csv files.
```
Comment (5tran-alexil) [inline development-guide.md:129]:
```suggestion
- `replace_files` is for the `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the .csv files.
```

File: development-guide.md [hunk 9]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
 
 - `update_files` is for `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the csv files.
```
Comment (5tran-alexil) [inline development-guide.md:131]:
```suggestion
- `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
```

File: development-guide.md [hunk 10]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
     - `unmodified_string` value is used to indicate columns in `update_files` where the values did not change.
 
+#### Capabilities
+This operation offers several additional features, as listed below:
+
+- Datatype Mappings: Supports adjusting partner data types for each Fivetran data type.
+- Max value for columns: Provides an option to specify the maximum value for each data type.
+
+#### AlterTable
+This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides details about these updates:
```
Comment (5tran-alexil) [inline development-guide.md:154]:
```suggestion
This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details about these updates:
```

File: development-guide.md [hunk 11]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in destination table.
```
Comment (5tran-alexil) [inline development-guide.md:155]:
```suggestion
- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in a destination table.
```

File: development-guide.md (hunk 6 above)
Comment (fivetran-satvikpatil) [reply development-guide.md]: I have done the changes, please take a look.

File: development-guide.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fivetran-satvikpatil) [reply development-guide.md]: I have changed the statement, please take a look.

**Review by 5tran-alexil — APPROVED** [review]
LGTM

File: development-guide.md [hunk 12]
```diff
@@ -61,13 +61,32 @@ The following are hard requirements to be able to deploy Partner code to Fivetra
...
 - Don't forget to handle new schemas/tables/columns per the information and user choices in `UpdateRequest#selection`.
 - Make sure you checkpoint at least once an hour. The more frequently you do it, the better.
 
-## Destination Guidelines
+### RPC Calls
+#### Schema
+This operation retrieves the customer's schemas, tables, and columns. It also includes an optional `selection_not_supported` field that indicates whether customers can select or deselect tables and columns within the Fivetran dashboard.
+
+#### Update
+This operation should retrieve data from the source. We provide a request using the `UpdateRequest` message, which includes the customer's state, credentials, and schema information. The response, delivered through the `UpdateResponse` message, should contain data records or other supported operations.
```
Comment (ediril) [inline development-guide.md:87]:
```suggestion
This operation should retrieve data from the source. We send a request using the `UpdateRequest` message, which includes the customer's state, credentials, and schema information. The response, streaming through the `UpdateResponse` message, can contain data records and other supported operations.
```

File: development-guide.md [hunk 13]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
 - This operation might be requested for a table that does not exist in the destination. In that case, it should NOT fail, simply ignore the request and return `success = true`.
 - `utc_delete_before` has millisecond precision.
 
-#### WriteBatchRequest
-- `replace_files` is for `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the csv files.
+#### WriteBatch
+This operation provides details about the batch files containing the retrieved data. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```
Comment (ediril) [inline development-guide.md:128]:
```suggestion
This operation provides details about the batch files containing the records to be pushed to the destination. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

File: development-guide.md [hunk 14]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
+- `replace_files` is for the `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
 
-- `update_files` is for `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the csv files.
+- `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
 
 - `delete_files` is for `hard delete` operation. Use primary key columns (or `_fivetran_id` system column for primary-keyless tables) to perform `DELETE FROM`.
 
+- `keys` is a map that provides a list of secret keys that can be used to decrypt batch files.
```
Comment (ediril) [inline development-guide.md:135]:
```suggestion
- `keys` is a map that provides a list of secret keys, one for each batch file, that can be used to decrypt them.
```

File: development-guide.md [hunk 15]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
     - `null_string` value is used to represent `NULL` value in all batch files.
     - `unmodified_string` value is used to indicate columns in `update_files` where the values did not change.
 
+#### Capabilities
+This operation offers several additional features, as listed below:
```
Comment (ediril) [inline development-guide.md:148]:
```suggestion
This operation offers the ability for the partner code to declare its choices for capabilities listed below:
```

File: development-guide.md [hunk 16]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+
+- Datatype Mappings: Supports adjusting partner data types for each Fivetran data type.
```
Comment (ediril) [inline development-guide.md:150]:
```suggestion
- Datatype Mappings: Provides the option to map destination data types to Fivetran data types.
```

File: development-guide.md [hunk 17]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+- Max value for columns: Provides an option to specify the maximum value for each data type.
+
+#### AlterTable
+This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details about these updates:
```
Comment (ediril) [inline development-guide.md:154]:
```suggestion
This operation is used to communicate changes to a table as specific update operations. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details:
```

**Review by ediril — CHANGES_REQUESTED** [review]

**Review by ediril — APPROVED** [review]

File: development-guide.md [hunk 18]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
+- `replace_files` is for the `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
 
-- `update_files` is for `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the csv files.
+- `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
```
Comment (fivetran-abdulsalam) [inline development-guide.md:131]: Now partners can also get parquet file as well, so this .CSV files doesn't make sense

File: development-guide.md [hunk 19]
```diff
@@ -120,6 +144,19 @@ Do not assume order of columns in the batch files. Always read the CSV file head
...
+This operation offers the ability for the partner code to declare its choices for capabilities listed below:
+
+- Datatype Mappings: Provides the option to map destination data types to Fivetran data types.
+- Max value for columns: Provides an option to specify the maximum value for each data type.
+
+#### AlterTable
+This operation is used to communicate changes to a table as specific update operations. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details:
+- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in a destination table.
+- Update Column type (`change_column_type`): This field provides information on updated type of a column in the source that needs to be reflected in a destination table.
+- Primary key updates (`updated_primary_keys`):  If the primary key has changed, this field lists all the columns used in the updated primary key.
```
Comment (fivetran-abdulsalam) [inline development-guide.md:157]: We have changed it to `update_primary_keys`
```suggestion
- Primary key updates (`update_primary_keys`):  If the primary key has changed, this field lists all the columns used in the updated primary key.
```

**Review by fivetran-abdulsalam — CHANGES_REQUESTED** [review]
I have requested some changes

File: development-guide.md (hunk 18 above)
Comment (ediril) [reply development-guide.md:131]: Good catch @fivetran-abdulsalam! We should change `CsvFileParams` to `file_params` which covers both CSV and Parquet

File: development-guide.md [hunk 20]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
+- `replace_files` is for the `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the batch files.
 
-- `update_files` is for `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the csv files.
+- `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `FileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the batch files.
```
Comment (fivetran-satvikpatil) [inline development-guide.md:131]: Note: I will merging the `CsvFileParams` and `ParquetFileParams` into one i.e. `FileParams`

File: development-guide.md [hunk 21]
```diff
@@ -105,13 +124,18 @@ This operation should report all columns in the destination table, including Fiv
...
 
 - `delete_files` is for `hard delete` operation. Use primary key columns (or `_fivetran_id` system column for primary-keyless tables) to perform `DELETE FROM`.
 
+- `keys` is a map that provides a list of secret keys, one for each batch file, that can be used to decrypt them.
+
+- `file_params` provides information about the file type and any configurations applied to it, such as encryption or compression.
+
 Also, Fivetran will deduplicate operations such that each primary key will show up only once in any of the operations
 
 Do not assume order of columns in the batch files. Always read the CSV file header to determine column order.
```
Comment (fivetran-abdulsalam) [inline development-guide.md:141]:
```suggestion
Do not assume order of columns in the batch files. Always read the columns of batch file to determine column order.
```

**Review by fivetran-abdulsalam — APPROVED** [review]
Approved with a nit comment

**Comment by fivetran-satvikpatil** [comment]
Related PR: https://github.com/fivetran/fivetran_sdk/pull/72

**Pull request closed without merging**